### PR TITLE
Feat update token seen at

### PR DESCRIPTION
**Pull request by nunogois**
https://linear.app/unleash/issue/2-448/update-last-seen-column-for-api-tokens

### CODE REVIEW

File: src/lib/db/api-token-store.ts [hunk 1]
```diff
@@ -200,7 +200,7 @@ export class ApiTokenStore implements IApiTokenStore {
         const now = new Date();
         try {
             await this.db(TABLE)
-                .whereIn('secrets', secrets)
+                .whereIn('secret', secrets)
```
Comment (gastonfournier) [inline src/lib/db/api-token-store.ts:203]: Is this a bug unrelated to this PR? Just curious

Comment (nunogois) [reply src/lib/db/api-token-store.ts:203]: Yep, this `markSeenAt` method was already implemented from a while ago, git says March 2021, except it was not being used AFAIK and had this typo. So I guess I'm following the approach that was originally thought of, with the `secrets string[]` parameter and all.

File: src/lib/services/api-token-service.ts [hunk 2]
```diff
@@ -46,15 +47,25 @@ export class ApiTokenService {
 
     private timer: NodeJS.Timeout;
 
+    private seenTimer: NodeJS.Timeout;
+
     private activeTokens: IApiToken[] = [];
 
+    private lastSeenSecrets: string[] = [];
```
Comment (gastonfournier) [inline src/lib/services/api-token-service.ts:54]: One problem with this approach could be memory consumption, cause we might end up with a large array in memory. I think it'd be better to have a Map structure and then use the keys to obtain the last seen secrets.

Comment (nunogois) [reply src/lib/services/api-token-service.ts:54]: The array gets cleared in https://github.com/Unleash/unleash/pull/2514/files/e31c6d6296e7b5317e86e502abbbe4ff158c0b76#diff-c7747d86059ec3422885e79411b8ef7323c0804df081c110268dd3250ace3696R102 - But we can also try a Map approach.

Comment (nunogois) [reply src/lib/services/api-token-service.ts:54]: I remembered that you may be thinking about having an unique array (don't have the same token more than once in the array). I agree with that and I think we could maybe use a Set here.

File: src/lib/services/api-token-service.ts [hunk 3]
```diff
@@ -79,6 +96,13 @@ export class ApiTokenService {
         }
     }
 
+    async updateLastSeen(): Promise<void> {
+        if (this.lastSeenSecrets.length > 0) {
+            await this.store.markSeenAt(this.lastSeenSecrets);
+            this.lastSeenSecrets = [];
```
Comment (gastonfournier) [inline src/lib/services/api-token-service.ts:102]: We could lose some last seen in between due to the interleaving of different operations. I don't know if there's an atomic way of doing this... you may know better, but I googled this:
```typescript
let toStore = [];
[this.lastSeenSecrets, toStore] = [toStore, this.lastSeenSecrets];
await this.store.markSeenAt(toStore);
```

Comment (nunogois) [reply src/lib/services/api-token-service.ts:102]: Interesting! That could work, thanks 👍

File: src/lib/services/api-token-service.ts (hunk 2 above)
Comment (gastonfournier) [reply src/lib/services/api-token-service.ts:54]: Yes, a set would work as well, the problem is with high-throughput applications, say 1000 req/s we'd be queuing 60K strings in memory every minute (which is 4.2Mb considering a token size of 70 characters). It's not much but if we're speaking of a single token, it'd be much better to have only 70 bytes in memory :D

Comment (nunogois) [reply src/lib/services/api-token-service.ts]: Thanks, tried addressing this in https://github.com/Unleash/unleash/pull/2514/commits/ed33001b04d69c056ed25827d2530a777e88ecf1 - Let me know what you think 👍 
I'll do the same change in the PAT PR.

File: src/lib/services/api-token-service.ts [hunk 4]
```diff
@@ -63,6 +74,12 @@ export class ApiTokenService {
             () => this.fetchActiveTokens(),
             minutesToMilliseconds(1),
         ).unref();
+        if (this.flagResolver.isEnabled('tokensLastSeen')) {
+            this.seenTimer = setInterval(
+                () => this.updateLastSeen(),
```
Comment (gastonfournier) [inline src/lib/services/api-token-service.ts:79]: Should we await here? Something like:
```suggestion
                () => await this.updateLastSeen(),
```
The concern is: if `updateLastSeen` executes asynchronously, some updateLastSeen might overlap over time (i.e. run 2 in parallel). It'd take quite a lot of time to happen with 1 minute between operations, but I've seen this before. 

I usually wait until the previous operation finishes before triggering the timer again. I'm not sure how this is done in node (or any best practice). Just wanted to touch base on this but it's not a blocker

**Review by gastonfournier — COMMENTED** [review]
This looks great IMO, just a small question but I might be overthinking it.

File: src/lib/services/api-token-service.ts (hunk 4 above)
Comment (nunogois) [reply src/lib/services/api-token-service.ts]: I don't think it works like this, as it still doesn't ensure sequentiality, but I may be wrong.
What do you think about an approach like https://github.com/Unleash/unleash/pull/2514/commits/47a1f0a8b8da574c23253076b1941e6231b516b6 inspired by: https://github.com/Unleash/unleash/blob/a992aca228c3bc0c17abcac22d027796db8059c0/src/lib/proxy/proxy-repository.ts#L91-L97

Comment (gastonfournier) [reply src/lib/services/api-token-service.ts]: Yes, I read a bit about the topic and your change is in line with my intuition: setTimeout is less risky (operationally speaking) cause there's no risk of executing multiple instances of the code concurrently. It does lag if the function takes longer, but that's exactly what I'd like to see in this particular use case: to execute less frequently in situations of high CPU load or other situations that may cause this function to execute slower.

Resources:
- https://blog.skay.dev/when-to-use-settimeout-vs-setinterval#conclusion
- https://stackoverflow.com/questions/729921/settimeout-or-setinterval

File: src/lib/db/api-token-store.ts (hunk 1 above)
Comment (ivarconr) [reply src/lib/db/api-token-store.ts:206]: nice catch! 🗡️

File: src/lib/services/api-token-service.ts [hunk 5]
```diff
@@ -92,6 +106,19 @@ export class ApiTokenService {
         }
     }
 
+    async updateLastSeen(): Promise<void> {
+        if (this.lastSeenSecrets.size > 0) {
+            let toStore = new Set<string>();
+            [this.lastSeenSecrets, toStore] = [toStore, this.lastSeenSecrets];
+            await this.store.markSeenAt([...toStore]);
```
Comment (ivarconr) [inline src/lib/services/api-token-service.ts:113]: This syntax makes me a bit nervous, as it requires the next line to be a spread operator. Remember that node processes are single threaded, so we can have two lines in a row and assume order (as long as they are not async). 

**How about:**

```suggestion
            let toStore = [...this.lastSeenSecrets];
            this.lastSeenSecrets = new Set<string>();
            await this.store.markSeenAt(toStore);
```

File: src/lib/services/api-token-service.ts [hunk 6]
```diff
@@ -52,10 +53,16 @@ export class ApiTokenService {
 
     private timer: NodeJS.Timeout;
 
+    private seenTimer: NodeJS.Timeout;
```
Comment (ivarconr) [inline src/lib/services/api-token-service.ts:56]: We could have made the precious timer in to an array, we only need to keep the reference to be able to clear them all.

Comment (nunogois) [reply src/lib/services/api-token-service.ts:56]: Hmm I'm not sure I understood this one, could you please clarify? 
I guess they could be a `NodeJS.Timeout[]` but if we want a single reference to each we would store them in `[0]` and `[1]` I guess? 🤔 Besides one of them is an interval and the other a timeout. 
Seems like decreased readability without much to gain, so I'm probably missing something.

**Pull request merged**